### PR TITLE
tools/cephfs-data-scan: support for multi-datapool

### PR DESCRIPTION
**Pull request by trociny**
Fixes: https://tracker.ceph.com/issues/58029
Signed-off-by: Mykola Golub <mykola.golub@clyso.com>

## Contribution Guidelines
- To sign and title your commits, please refer to [Submitting Patches to Ceph](https://github.com/ceph/ceph/blob/main/SubmittingPatches.rst).

- If you are submitting a fix for a stable branch (e.g. "pacific"), please refer to [Submitting Patches to Ceph - Backports](https://github.com/ceph/ceph/blob/master/SubmittingPatches-backports.rst) for the proper workflow.

## Checklist
- Tracker (select at least one)
  - [ ] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [ ] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [ ] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [ ] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
</details>

### CODE REVIEW

**Comment by vshankar** [comment]
@trociny Changes looks fine. We should add some tests though (in qa/tasks/cephfs/test_data_scan.py). Currently the tests assume a single data pool. We can extend that...

**Comment by vshankar** [comment]
And a note in doc/cephfs/disaster-recovery-experts.rst regarding the same would be good.

**Comment by vshankar** [comment]
@trociny ping?

**Comment by trociny** [comment]
> @trociny ping?

@vshankar Sorry, I have started working on your suggestions. Though being distracted these days and this thing is not of highest priority I still need some time to extend the tests.

**Comment by vshankar** [comment]
> > @trociny ping?
> 
> @vshankar Sorry, I have started working on your suggestions. Though being distracted these days and this thing is not of the highest priority I still need some time to extend the tests.

No problem - I was gathering some PRs to run through tests. Ping me when this is ready.

**Comment by trociny** [comment]
@vshankar I think I have fixed the issue and it is ready for review now.

I was surprised to learn that we store the "layout" attribute in the the first inode object in the "extra" data pool. I was expecting to find it in the object from the main data pool. Is there a special reason for this?

Anyway, I had to modify `ClsCephFSClient::accumulate_inode_metadata` to also store the inode object pool id in an additional attribute.  I updated the cls method to do this, which means that an updated tool will not work as expected for an outdated cluster (for multi-data pool case). If we think it is important to support this case I can change it to just call `setxattr`, leaving the `accumulate_inode_metadata` cls method unchanged.

**Comment by vshankar** [comment]
> @vshankar I think I have fixed the issue and it is ready for review now.
> 
> I was surprised to learn that we store the "layout" attribute in the the first inode object in the "extra" data pool. I was expecting to find it in the object from the main data pool. Is there a special reason for this?

The object in the default data pool _always_ holds the backtrace - layouts are stored in the data pool customized (if any) via file/dir layouts.

I'll have a look at your change today.

**Comment by vshankar** [comment]
> Anyway, I had to modify `ClsCephFSClient::accumulate_inode_metadata` to also store the inode object pool id in an additional attribute. I updated the cls method to do this, which means that an updated tool will not work as expected for an outdated cluster (for multi-data pool case). If we think it is important to support this case I can change it to just call `setxattr`, leaving the `accumulate_inode_metadata` cls method unchanged.

IMO, we need to make the tool work on older clusters. I don't think that would break anything. @batrick WDYT?

File: src/cls/cephfs/cls_cephfs_client.cc [hunk 1]
```diff
@@ -66,6 +70,8 @@ int ClsCephFSClient::delete_inode_accumulate_result(
   op.rmxattr(XATTR_CEILING);
   op.rmxattr(XATTR_MAX_SIZE);
   op.rmxattr(XATTR_MAX_MTIME);
+  op.rmxattr(XATTR_POOL_ID);
+  op.set_op_flags2(librados::OP_FAILOK);
```
[the remark this answers is not in the document]
Comment (trociny) [reply src/cls/cephfs/cls_cephfs_client.cc:74]: Yes, I believe so.

https://github.com/ceph/ceph/blob/main/src/librados/librados_cxx.cc#L118

**Comment by trociny** [comment]
> IMO, we need to make the tool work on older clusters. I don't think that would break anything. 

I updated the PR to use `setxattr` instead of modifying the class method.

**Comment by vshankar** [comment]
> > IMO, we need to make the tool work on older clusters. I don't think that would break anything.
> 
> I updated the PR to use `setxattr` instead of modifying the class method.

I'll have a look today.

**Review by vshankar — COMMENTED** [review]
Looks fine. I'll put this to test with the next batch of PRs.

**Comment by vshankar** [comment]
jenkins retest this please

**Comment by vshankar** [comment]
Test ran successfully with this change - https://pulpito.ceph.com/vshankar-2023-02-24_02:11:45-fs-wip-vshankar-testing-20230222.025426-testing-default-smithi/

Failures are not related to this PR.

**Comment by batrick** [comment]
jenkins test make check

File: qa/tasks/cephfs/cephfs_test_case.py [hunk 2]
```diff
@@ -161,9 +161,10 @@ def setUp(self):
 
             # In case some test messed with auth caps, reset them
             for client_id in client_mount_ids:
-                cmd = ['auth', 'caps', f'client.{client_id}', 'mon','allow r',
-                       'osd', f'allow rw pool={self.fs.get_data_pool_name()}',
-                       'mds', 'allow']
+                cmd = ['auth', 'caps', f'client.{client_id}',
+                       'mds', f'allow rwp fsname={self.fs.name}',
```
Comment (vshankar) [inline qa/tasks/cephfs/cephfs_test_case.py:165]: @trociny Looks like we have a couple of issues here:

This line is failing to be parsed resulting and results in the following error
```                                              
2023-03-01T04:31:47.882 INFO:teuthology.orchestra.run.smithi008.stderr:Error EINVAL: mds capability parse failed, stopped at 'fsname=cephfs' of 'allow rwp fsname=cephfs'                                         
2023-03-01T04:31:47.904 DEBUG:teuthology.orchestra.run:got remote process result: 22
```

Failed job: https://pulpito.ceph.com/vshankar-2023-03-01_02:28:49-fs-wip-vshankar-testing-20230228.105516-testing-default-smithi/7190048/

I didn't debug this in detail as I'm on PTO today.

Another thing is we need to include the 's' flag (after rwp) for the user to be able to create and delete snapshots. See: https://docs.ceph.com/en/latest/cephfs/client-auth/#snapshot-restriction-the-s-flag

Comment (trociny) [reply qa/tasks/cephfs/cephfs_test_case.py:165]: As I understand the issue is that we are running this command against nautilus cluster and it does not support the `fsname` keyword yet.  Actually, I had to fix the osd caps only, and `'allow rw tag cephfs data={self.fs.name}'` seems to work on nautilus too. The mds and mon changes were just "improvement" to have it more strict. I will try just reverting them back and see if it helps.

As for the missing 's' flag, in the original version we just had `'mds', 'allow'`, which I suppose includes all flags, right? Then I will just revert to that version.

**Comment by batrick** [comment]
> IMO, we need to make the tool work on older clusters. I don't think that would break anything. @batrick WDYT?

I think it's okay to not work on older clusters if the tool gives an appropriate error message when trying to recover from multiple data pools.

**Comment by trociny** [comment]
I believe with my last changes it should work on the old cluster. I just have not tested it yet.

File: qa/tasks/cephfs/cephfs_test_case.py (hunk 2 above)
Comment (vshankar) [reply qa/tasks/cephfs/cephfs_test_case.py]: > As I understand the issue is that we are running this command against nautilus cluster and it does not support the `fsname` keyword yet. Actually, I had to fix the osd caps only, and `'allow rw tag cephfs data={self.fs.name}'` seems to work on nautilus too. The mds and mon changes were just "improvement" to have it more strict. I will try just reverting them back and see if it helps.
> 
> As for the missing 's' flag, in the original version we just had `'mds', 'allow'`, which I suppose includes all flags, right? Then I will just revert to that version.

Right.

**Comment by vshankar** [comment]
jenkins retest this please

**Comment by vshankar** [comment]
> > IMO, we need to make the tool work on older clusters. I don't think that would break anything. @batrick WDYT?
> 
> I think it's okay to not work on older clusters if the tool gives an appropriate error message when trying to recover from multiple data pools.

The changes to support this are pretty straight-forward. I think it would be beneficial to have this in place.

**Comment by vshankar** [comment]
jenkins test make check

**Comment by trociny** [comment]
Just FYI, I scheduled some teuthology jobs from fs suite, `--filter data-scan` [1] and `--filter fs/upgrade` [2] and there were no failures.

[1] https://pulpito.ceph.com/trociny-2023-03-02_12:40:33-fs-wip-mgolub-testing-distro-default-smithi/
[2] http://pulpito.front.sepia.ceph.com/trociny-2023-03-02_10:27:59-fs-wip-mgolub-testing-distro-default-smithi/

**Review by vshankar — APPROVED** [review]

**Pull request merged**